ProcessList.c

```diff
@@ -194,308 +191,142 @@ void ProcessList_remove(ProcessList* this, const Process* p) {
-         // use a ppid of 0. As that PID can't exist, we can skip searching for it.
-         if (!ppid)
-            r = 0;
-
-         while (l < r) {
-            int c = (l + r) / 2;
-            pid_t pid = ((Process*)Vector_get(this->processes, c))->pid;
-            if (ppid == pid) {
-               break;
-            } else if (ppid < pid) {
-               r = c;
-            } else {
-               l = c + 1;
-            }
-         }
+      // If PID corresponds with PPID (e.g. "kernel_task" (PID:0, PPID:0)
+      // on Mac OS X 10.11.6) regard this process as root.
+      if (process->pid == ppid) {
+         process->isRoot = true;
+         continue;
+      }
 
-         // If parent not found, then construct the tree with this node as root
-         if (l >= r) {
-            process = (Process*)Vector_take(this->processes, i);
-            process->indent = 0;
-            process->tree_depth = 0;
-            process->tree_left = node_counter++;
-            process->tree_index = node_index++;
-            Vector_add(this->processes2, process);
-            Hashtable_put(this->displayTreeSet, process->tree_index, process);
-            ProcessList_buildTreeBranch(this, process->pid, 0, 0, direction, process->showChildren, &node_counter, &node_index);
-            process->tree_right = node_counter++;
-            break;
-         }
+      // On Linux both the init process (pid 1) and the root UMH kernel thread (pid 2)
+      // use a ppid of 0. As that PID can't exist, we can skip searching for it.
+      if (!ppid) {
+         process->isRoot = true;
+         continue;
       }
 
-      // There should be no loop in the process tree
-      assert(i < size);
+      // We don't know about its parent for whatever reason
+      if (ProcessList_findProcess(this, ppid) == NULL)
+         process->isRoot = true;
+   }
+
+   // Sort by known parent PID (roots first), then PID
+   Vector_quickSortCustomCompare(this->processes, compareProcessByKnownParentThenNatural);
+
+   // Find all processes whose parent is not visible
+   for (int i = 0; i < vsize; i++) {
+      Process* process = (Process*)Vector_get(this->processes, i);
+
+      // If parent not found, then construct the tree with this node as root
+      if (process->isRoot) {
+         process = (Process*)Vector_get(this->processes, i);
+         process->indent = 0;
+         process->tree_depth = 0;
+         Vector_add(this->displayList, process);
+         ProcessList_buildTreeBranch(this, process->pid, 0, 0, process->showChildren);
+         continue;
+      }
    }
 
-   // Swap listings around
-   Vector* t = this->processes;
-   this->processes = this->processes2;
-   this->processes2 = t;
+   this->needsSort = false;
 
    // Check consistency of the built structures
-   assert(Vector_size(this->processes) == vsize); (void)vsize;
-   assert(Vector_size(this->processes2) == 0);
+   assert(Vector_size(this->displayList) == vsize); (void)vsize;
```

Comment:
This assert is more interesting: it checks that the number of processes in the display list matches the full number of processes. On the one hand, it highlights that the "display list" here actually contains all processes, not only ones that are actually displayed according to out filters.
On the other hand, while a process cannot get into the list more than once, this assert can trigger if the current snapshot contains a race-caused parent loop.